### PR TITLE
HEEDLS-258 Back end for course completion summary

### PR DESCRIPTION
**Pull request by benjimarshall**
## Changes
Make a CourseCompletionService and data model for the information required by the course completion page

## Testing
Added unit tests for this service

### CODE REVIEW

File: DigitalLearningSolutions.Data.Tests/Services/CourseCompletionServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                13,
+                9,
+                13
+            ));
+        }
+
+        [Test]
+        public void Get_course_completion_of_course_should_when_all_tutorials_status_0()
+        {
+            // CustomisationTutorials.Status = 0 entries where CustomisationID = 26696
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data.Tests/Services/CourseCompletionServiceTests.cs:55]: What exactly does this comment mean? 🙂

File: DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
+﻿namespace DigitalLearningSolutions.Data.Models.CourseCompletion
+{
+    using System;
+
+    public class CourseCompletion
+    {
+        public int Id { get; }
+        public string ApplicationName { get; }
+        public string CustomisationName { get; }
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs:9]: Should the two fields above be combined to a Title fields, as we have done for other models? Or is there a reason to keep them separate? 🙂

File: DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        public bool IncludeCertification { get; }
+        public DateTime? Completed { get; }
+        public DateTime? Evaluated { get; }
+        public int MaxPostLearningAssessmentAttempts { get; }
+        public bool IsAssessed { get; }
+        public int PostLearningAssessmentPassThreshold { get; }
+        public int DiagnosticAssessmentCompletionThreshold { get; }
+        public int TutorialsCompletionThreshold { get; }
+        public int? DiagnosticScore { get; }
+        public int? LearningDone { get; }
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs]: What does this field represent? 🙂

File: DigitalLearningSolutions.Data/Services/CourseCompletionService.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                         INNER JOIN Applications
+                         ON Customisations.ApplicationID = Applications.ApplicationID
+
+                         INNER JOIN Sections
+                         ON Applications.ApplicationID = Sections.ApplicationID
+                   WHERE Customisations.CustomisationID = @customisationId
+                         AND Sections.ArchivedDate IS NULL
+                   GROUP BY Customisations.CustomisationID
+                  ),
+                         PostLearningPasses AS (
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/CourseCompletionService.cs:37]: Indentation a bit off? 🙂

**Review by FridaTveit — COMMENTED** [review]
A few comments, otherwise looks good 🙂

**Comment by benjimarshall** [comment]
That previous commit was adding a missing field from the service (the number of diagnostic attempts made) and adding a few more tests. I forgot a few bits whilst finishing off before Christmas, and those are the bits I realised later on that I needed to do 🙂

File: DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs (hunk 3 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs]: `DiagnosticAssessmentCompletionThreshold` corresponds to `DiagCompletionThreshold` from the old stored procedure, and will be used to produce text like "To complete this course, you must achieve {DiagnosticAssessmentCompletionThreshold}% in the diagnostic assessment" 🙂

Comment (FridaTveit) [reply DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs]: Sorry I meant the line above `LearningDone` 🙂

File: DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs (hunk 2 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs]: No reason, I had just forgotten about that! Changed in 2cbdd33 🙂

File: DigitalLearningSolutions.Data.Tests/Services/CourseCompletionServiceTests.cs (hunk 1 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data.Tests/Services/CourseCompletionServiceTests.cs]: That's not very clear, sorry! Changed to a hopefully cleared comment in 2cbdd33 to:

`CustomisationTutorials.Status = 0 for all CustomisationTutorial records where CustomisationID = 26696`

File: DigitalLearningSolutions.Data/Services/CourseCompletionService.cs (hunk 4 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/CourseCompletionService.cs]: Yep! Fixed in 2cbdd33 🙂

File: DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs (hunk 3 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs]: Ah this is the name of the same field from the old stored procedure, it essentially is the percentage of tutorials completed. I could change the name to something clearer, maybe `TutorialsCompleted`? 🙂

Comment (FridaTveit) [reply DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs]: Yes I think even `PercentageTutorialsCompleted`, to be completely clear 🙂

File: DigitalLearningSolutions.Data/Services/CourseCompletionService.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                         SELECT AssessAttempts.ProgressID,
+                               MAX(CONVERT(tinyint, Status)) AS HasPassed
+                          FROM AssessAttempts
+                         WHERE AssessAttempts.CustomisationID = @customisationId
+                               AND AssessAttempts.CandidateID = @candidateId
+                         GROUP BY AssessAttempts.ProgressID, AssessAttempts.SectionNumber
+                    ) AS PostLearningPassStatus
+                   GROUP BY PostLearningPassStatus.ProgressID
+                  ),
+                         LearningDone AS (
```
Comment (FridaTveit) [inline DigitalLearningSolutions.Data/Services/CourseCompletionService.cs:50]: Indentation also off here 🙂

**Review by FridaTveit — COMMENTED** [review]
One small thing, otherwise looks good 🙂

File: DigitalLearningSolutions.Data/Services/CourseCompletionService.cs (hunk 5 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Services/CourseCompletionService.cs]: Yep, thanks! Fixed in 64314db 🙂

File: DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs (hunk 3 above)
Comment (benjimarshall) [reply DigitalLearningSolutions.Data/Models/CourseCompletion/CourseCompletion.cs]: Renamed in 64314db 🙂

**Review by FridaTveit — APPROVED** [review]
Looks good 🙂

**Pull request merged**